### ISSUE TITLE
Setting up centerline and network model nodes for centerline extraction (Python scripting)

### ISSUE THREAD

**Issue by perecanals** [vmtk/SlicerExtension-VMTK#23]
Hi!

I am trying to automate segmentation + centerline extraction from a binary mask of a vascular structure and so far I have managed to successfully get the segmentation, setup the `extractCenterlineWidget` (including setting the `segmentationNode` as the input surface node for the Extract Centerline module) and extract endpoints via the Auto-detect function (`onAutoDetectEndPoints()`). At this point I am stuck in setting up the centerline module and the network module nodes before instancing the `onApplyButton()` method, which, as I understand it, should extract the centerline model and network model if nodes are correctly set up in the `._parameterNode` of the widget (is this correct?).

I have gone through the methods of `._parameterNode` and I do not seem to identify how to correctly set up the `CenterlineModel` or the `NetworkModel` node to extract them via the `onApplyButton()` method, how should I do this? Also, how do I know the node types that these should be?

I think I have managed to obtain the vtkPolyData objects of the centerline and the network models through the `.logic` methods, but I would like to visualize it in the GUI through Python scripting.

I can provide code and data if needed.

Thank you,

**Comment by perecanals** [comment]
Nevermind, solved it!

**Comment by lassoan** [comment]
Thanks for the update.

Is there anything specific that you learned that could be useful to know for someone who wants to do something similar in the future?

**Comment by perecanals** [comment]
Hi,

Yes, sorry. I have been getting more familiar with how to navigate through the API wiki. I explored the possible node classes that I thought could be compatible with the centerline and network models in the [docs](https://apidocs.slicer.org/master/annotated.html) and I used `centerlineModelNode = slicer.mrmlScene.AddNewNodeByClass()` to set up the centerline model node. I think I chose `vtkMRMLMarkupsCurveNode` as a potential class for the centerline model node. Then I linked it to the `extractCenterlineWidget` (`extractCenterlineWidget = slicer.modules.extractcenterline.widgetRepresentation().self()`) by `extractCenterlineWidget._parameterNode.SetNodeReferenceID("CenterlineModel", centerlineModelNode.GetID())`. I then run the apply button and an error popped up indicating that the chosen class for the `centerlineModelNode` did not have some method or attribute, which then I searched for in the docs, and found that the only class with that method was the `vtkMRMLModelNode`, so I repeated the operation changing the node classes in the definition of the `centerlineModeNode`.

So basically, trial and error. The main problem I had was that I had no idea of how to know the class that a certain node should have, and also wasn't sure on how to associate a certain node to the `_parameterNode` of a widget class. 

I hope this helps, and if it's not very clear or you have any doubts, I'll be happy to help.

**Comment by zhuyinheng** [comment]
> Hi,
> 
> Yes, sorry. I have been getting more familiar with how to navigate through the API wiki. I explored the possible node classes that I thought could be compatible with the centerline and network models in the [docs](https://apidocs.slicer.org/master/annotated.html) and I used `centerlineModelNode = slicer.mrmlScene.AddNewNodeByClass()` to set up the centerline model node. I think I chose `vtkMRMLMarkupsCurveNode` as a potential class for the centerline model node. Then I linked it to the `extractCenterlineWidget` (`extractCenterlineWidget = slicer.modules.extractcenterline.widgetRepresentation().self()`) by `extractCenterlineWidget._parameterNode.SetNodeReferenceID("CenterlineModel", centerlineModelNode.GetID())`. I then run the apply button and an error popped up indicating that the chosen class for the `centerlineModelNode` did not have some method or attribute, which then I searched for in the docs, and found that the only class with that method was the `vtkMRMLModelNode`, so I repeated the operation changing the node classes in the definition of the `centerlineModeNode`.
> 
> So basically, trial and error. The main problem I had was that I had no idea of how to know the class that a certain node should have, and also wasn't sure on how to associate a certain node to the `_parameterNode` of a widget class.
> 
> I hope this helps, and if it's not very clear or you have any doubts, I'll be happy to help.

Could you please share the script that you have been working on(using VMTK slicer, via widget instead of logic). My email address is zhuyh19@mails.tsinghua.edu.cn
